sqld/src/database/write_proxy/replication.rs

```diff
@@ -207,43 +200,149 @@ fn free_page_header(h: *const PgHdr) {
     }
 }
 
+#[repr(C)]
+#[derive(Debug, Pod, Zeroable, Clone, Copy)]
+struct WalIndexMeta {
+    /// This is the anticipated next frame index to request
+    pre_commit_index: u64,
+    /// After we have written the frames back to the wal, we set this value to the same value as
+    /// pre_commit_index
+    /// On startup we check this value against the pre-commit value to check for consistency
+    post_commit_index: u64,
+    /// Generation Uuid
+    generation_id: u128,
+    /// Uuid of the database this instance is a replica of
+    database_id: u128,
+}
+
+#[derive(Debug, thiserror::Error)]
+enum MetaMergeError {
+    #[error("Replica is ahead of primary")]
+    Lagging,
+    #[error("Trying to replicate incompatible databases")]
+    DbIncompatible,
+    #[error("{0}")]
+    Other(#[from] anyhow::Error),
+}
+
+impl WalIndexMeta {
+    fn read(meta_file: &File) -> anyhow::Result<Option<Self>> {
+        let mut buf = [0; size_of::<WalIndexMeta>()];
+        let meta = match meta_file.read_exact_at(&mut buf, 0) {
+            Ok(()) => {
+                meta_file.read_exact_at(&mut buf, 0)?;
+                let meta = *try_from_bytes(&buf)
+                    .map_err(|_| anyhow::anyhow!("invalid index meta file"))?;
+                Some(meta)
+            }
+            Err(e) if e.kind() == ErrorKind::UnexpectedEof => None,
+            Err(e) => Err(e)?,
+        };
+
+        Ok(meta)
+    }
+
+    /// attempts to merge two meta files.
+    fn merge_from_hello(mut self, hello: HelloResponse) -> Result<Self, MetaMergeError> {
+        let hello_db_id = Uuid::from_str(&hello.database_id)
+            .context("invalid database id from primary")?
+            .as_u128();
+        let hello_gen_id = Uuid::from_str(&hello.generation_id)
+            .context("invalid generation id from primary")?
+            .as_u128();
+
+        if hello_db_id == self.database_id {
```

Comment:
Shouldn't it be != instead of ==?

Reply:
oops yeah I changed this to explicitly test test this error being triggered and forgot to change it back

Reply:
oh wait actually it's solved later